tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,279 @@
+                    return False
+                return True  # if no exception, just return
+            except Exception:
+                sleep_sec = uniform(1, 4)  # random float 1 to 4
+                self.logger.warning(
+                    'Caught exception sending to Splunk; '
+                    'retry in %s seconds', sleep_sec
+                )
+                sleep(sleep_sec)
+        self.logger.error(
+            'ERROR - Could not POST to Splunk after %d tries.', max_attempts
+        )
+        return False
+
+    def _send_splunk(self, payload):
+        """
+        Perform the actual data send to Splunk HEC for one log entry
+
+        :param payload: the JSON-encoded payload to send to Splunk as
+          event data
+        :type payload: str
+        """
+        url = self.config['splunk_hec_url']
+        self.logger.debug('Send to Splunk (%s): %s', url, payload)
+        try:
+            r = requests.post(
+                url,
+                headers={
+                    'Authorization': 'Splunk %s' % self.config[
+                        'splunk_hec_token'
+                    ]
+                },
+                data=payload
+            )
+        except Exception:
+            self.logger.error('Exception during Splunk POST to %s of %s',
+                              url, payload, exc_info=True)
+            raise
+        self.logger.debug(
+            'Splunk POST got response code %s HEADERS=%s BODY: %s',
+            r.status_code, r.headers, r.text
+        )
+        if r.status_code not in [200, 201, 202]:
+            self.logger.error(
+                'Splunk POST returned non-20x response: %s HEADERS=%s BODY: %s',
+                r.status_code, r.headers, r.text
+            )
+            raise RuntimeError('POST returned %s' % r.status_code)
+        try:
+            j = r.json()
+        except Exception:
+            j = {'text': r.text}
+        if j['text'].lower() != 'success':
+            self.logger.error(
+                'Splunk POST returned non-success response: %s', j
+            )
+            raise RuntimeError('POST returned non-success response: %s' % j)
+
+    @staticmethod
+    def tags_for_resource(res):
+        """
+        Return the tags for a given resource, or an empty dict if they
+        can't be found.
+
+        :param res: c7n resource info
+        :type res: dict
+        :return: dict of tags
+        """
+        try:
+            t = res.get('Tags', res.get('tags', {}))
+            if t is None:
+                return {}
+            # it's a dict, return it right away
+            if isinstance(t, type({})):
+                return t
+            # if it's not a dict or list, unknown, return empty dict
+            if not isinstance(t, type([])):
+                return {}
+            # it's a list
+            tags = {}
+            for item in t:
+                if 'Key' in item and 'Value' in item:
```

Comment:
is there any resource you've encountered where tags is not a list with Key/Value?

Reply:
Once again, this may be a relic of older code. I can pull some of our ``resources.json.gz`` files from S3 and do a bit of analysis on them, since I assume those will accurately reflect the data that's being sent to mailer...